shell/platform/android/io/flutter/plugin/editing/ListenableEditingState.java

```diff
@@ -179,6 +257,54 @@ public SpannableStringBuilder replace(
       Log.e(TAG, "editing state should not be changed in a listener callback");
     }
 
+    Log.e(
+        "DELTAS",
+        "replace(" + start + ", " + end + ", " + tb + ", " + tbstart + ", " + tbend + ")");
+
+    final boolean isDeletionGreaterThanOne = end - (start + tbend) > 1;
+    final boolean isCalledFromDelete = tb == "" && tbstart == 0 && tbstart == tbend;
+
+    final boolean isReplacedByShorter = isDeletionGreaterThanOne && (tbend - tbstart < end - start);
+    final boolean isReplacedByLonger = tbend - tbstart > end - start;
+    final boolean isReplacedBySame = tbend - tbstart == end - start;
+
+    // Is deleting/inserting at the end of a composing region.
+    final boolean isDeletingInsideComposingRegion = !isReplacedByShorter && start + tbend < end;
+    final boolean isInsertingInsideComposingRegion = start + tbend > end;
+
+    // To consider the cases when autocorrect increases the length of the text being composed by
+    // one, but changes more than one character.
+    final boolean isOriginalComposingRegionTextChanged =
+        (isCalledFromDelete || isDeletingInsideComposingRegion || isReplacedByShorter)
+            || !toString()
+                .subSequence(start, end)
+                .equals(tb.toString().subSequence(tbstart, end - start));
+
+    final boolean isEqual =
+        toString().subSequence(start, end).equals(tb.toString().subSequence(tbstart, tbend));
+
+    // A replacement means the original composing region has changed, anything else will be
+    // considered an insertion.
+    final boolean isReplaced =
+        isOriginalComposingRegionTextChanged
+            && (isReplacedByLonger || isReplacedBySame || isReplacedByShorter);
+
+    if (isEqual) {
+      Log.e("DELTAS", "EQUALITY");
+      setDeltas(toString(), "", "EQUALITY", -1, -1);
+    } else if (isCalledFromDelete || isDeletingInsideComposingRegion) {
+      Log.e("DELTAS", "DELETION");
+      setDeltas(
+          toString(), toString().subSequence(start + tbend, end).toString(), "DELETION", end, end);
+    } else if ((start == end || isInsertingInsideComposingRegion)
+        && !isOriginalComposingRegionTextChanged) {
+      Log.e("DELTAS", "INSERTION");
+      setDeltas(toString(), tb.subSequence(end - start, tbend).toString(), "INSERTION", end, end);
+    } else if (isReplaced) {
+      Log.e("DELTAS", "REPLACEMENT");
+      setDeltas(toString(), tb.subSequence(tbstart, tbend).toString(), "REPLACEMENT", start, end);
+    }
```

Comment:
Is it ever possible that `setDeltas` won't be called here?  It seems like bugs could arise from that, like sending old deltas data in updateEditingStateWithDelta.

Reply:
And is it ever possible that some crazy change could happen that wouldn't be supported by a single TextEditingDelta?  Maybe Android is massaging these types of changes into a single call to `replace` for you.  Like what if you had two insertions at once?

```
"bbb" => "cbbbc"
```

Is there any way we can test something like that?  I guess Android would just replace the entire "bbb" string rather than trying to do an insertion at the beginning and another at the end.

Reply:
I think that might be possible not to hit `setDeltas` we definitely need to make sure to cover all the cases here. Yeah in the case you mention android would replace the entire string with `"cbbbc"`. I think in this case all we can say is what the platform gives us, and in Android's case it is that the entire word was replaced.